### DISCUSSION TITLE
생성자 체이닝이 무엇이고 왜 필요한지?

### DISCUSSION THREAD

**Discussion by GIVEN53** [2024-woowacourse-study/level-interview#22] [Java]
.

**Comment by alstn113** [comment]
생성자 체이닝은 자바에서 한 생성자가 다른 생성자를 호출하는 것을 말한다.
즉, 여러 개의 생성자가 있을 때 한 생성자에서 다른 생성자를 호출하여 초기화 작업을 수행하는 것이다.

생성자 체이닝은 다음과 같은 이유로 필요하다.
1. 코드 중복 최소화: 여러 개의 생성자가 있을 때, 각 생성자마다 초기화 작업을 반복하지 않고 한 곳에서 초기화 로직을 구현할 수 있음. 
2. 초기화 코드 재활용: 특정 생성자에서 초기화된 멤버 변수를 다른 생성자에서도 활용할 수 있다. 일관성 있는 객체 생성을 보장한다.
3. 유연한 객체 생성: 생성자 체이닝을 통해 다양한 초기화 방법을 제공할 수 있다.

**Comment by GIVEN53** [comment]
생성자 체이닝은 한 생성자에서 this 또는 super 키워드로 다른 생성자를 호출하는 것이다.
하나의 주생성자를 지정하고 주생성자에서만 필드를 초기화한다. 그 외의 부생성자는 주생성자를 호출하도록 구성한다.

### bad case
```java
public class Person {
    private final String name;
    private final int age;

    public Person(int age) {
        this.name = "프린";
        this.age = age;
    }

    public Person(String name, int age) {
        this.name = name;
        this.age = age;
    }
}
```

### good case
```java
public class Person {
    private final String name;
    private final int age;

    public Person(int age) {
        this("프린", age);
    }

    public Person(String name, int age) {
        this.name = name;
        this.age = age;
    }
}
```

### 장점
1. 주생성자에서만 필드를 초기화하기 때문에 도메인 유효성 검증을 주생성자에 집중할 수 있다.
2. 초기화 로직을 주생성자 한 곳에서 관리하여 코드 중복을 최소화하고 유지보수성을 높인다.
3. 외부에 다양한 생성자를 노출하여 유연하게 객체를 생성할 수 있도록 한다

**Comment by nayonsoso** [comment]
생성자 체이닝은 하나의 생성자에서 다른 생성자를 호출하도록 하는 방식입니다. 생성자 체이닝을 사용하면 중복되는 초기화 코드를 줄일 수 있다는 장점이 있습니다. 

생성자 체이닝을 효과적으로 하기 위해서 주생성자와 부생성자의 개념을 사용하면 좋습니다. 주생성자란, 필드 수와 일치하는 인자를 받아 모든 필드를 초기화하는 생성자를 말합니다. 부생성자는 일부 필드만 초기화하거나, 여러가지 생성 방식을 제공하여 주 생성자를 보조하는 역할을 합니다.

부생성자가 주생성자를 호출하게 하고, 주생성자를 모든 부생성자 뒤에 둔다면 가독성과 유지보수성을 향상시킬 수 있습니다.

**Comment by GIVEN53** [comment]
다른 생성자 호출
필드 초기화 주생성자
유효성 검증 집중
중복 제거, 유지보수성
유연한 객체 생성
